crates/client/sync/src/commitments/lib.rs

```diff
@@ -146,109 +162,130 @@ where
 /// # Returns
 ///
 /// The updated state root as a `Felt252Wrapper`.
-pub async fn update_state_root<B: BlockT>(
+pub fn update_state_root<B: BlockT>(
     csd: CommitmentStateDiff,
-    bonsai_dbs: BonsaiDbs<B>,
-) -> anyhow::Result<Felt252Wrapper> {
-    let arc_csd = Arc::new(csd);
-    let arc_bonsai_dbs = Arc::new(bonsai_dbs);
-
-    let contract_trie_root = contract_trie_root(Arc::clone(&arc_csd), Arc::clone(&arc_bonsai_dbs)).await?;
-
-    let class_trie_root = class_trie_root(Arc::clone(&arc_csd), Arc::clone(&arc_bonsai_dbs))?;
-
-    let state_root = calculate_state_root::<PoseidonHasher>(contract_trie_root, class_trie_root);
-
-    Ok(state_root)
+    overrides: Arc<OverrideHandle<Block<Header<u32, BlakeTwo256>, OpaqueExtrinsic>>>,
+    bonsai_contract: Arc<Mutex<BonsaiStorage<BasicId, BonsaiDb<B>, Pedersen>>>,
+    bonsai_class: Arc<Mutex<BonsaiStorage<BasicId, BonsaiDb<B>, Poseidon>>>,
+    block_number: u64,
+    substrate_block_hash: Option<H256>,
+) -> Felt252Wrapper {
+    // Update contract and its storage tries
+    let contract_trie_root =
+        contract_trie_root(&csd, overrides, bonsai_contract.lock().unwrap(), block_number, substrate_block_hash)
```

Comment:
pass the mutex as a parameter in order to lock for the shortest possible time in the child function

Reply:
same as above, this is not necessary in this case since the mutex is not used anywhere else (except maybe in RPC calls), and we definitely do not want any mutation to occur while we are computing the state root.